Enable cert updates

sevenmachines opened the pull request:
When we change the certificate in our cloud formation config we want a way to update the certificates on instances, and to set these certificates on any https load balancers on the stack. 
This patch creates a fab_task update_certs() to carry out both of these tasks, first removing/adding the certificates using IAM, then using a simple ELB class to set_ssl_certificates() on any https listeners on the stacks load_balancers

ashb commented on bootstrap_cfn/elb.py:
```diff
@@ -0,0 +1,75 @@
+                # necessarily a problem
+                logging.info("ELB::set_ssl_certificates: "
+                             "No load balancers found in stack, "
+                             "no certificate updates needed")
```
This probably be a warning or an error - the user has requested something to be updated that couldn't be. This also means we should likely return a non-0 exit code.

sevenmachines replied:
I was going to throw an exception here to allow the caller to deal with there being no elbs or ignoring, exiting seems a little strict?

ashb commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -317,3 +319,26 @@ def rsync():
+    # Set the certificates on ELB's
+    if 'elb' in cfn_config.data:
+        elb = ELB(aws_config)
+        elb.set_ssl_certificates(cfn_config.ssl(), stack_name)
```
Similar thing about the `if`s here - we should check the both exist to start with and error if they don't.

niallcreech replied:
If the certs dont exist then it should be a non-zero exit, but elb's aren't required to exist, so handling the exception when they dont seems ok to do.

sevenmachines replied:
Trying to update certs when there not defined in the config warrants a non-zero exit right enough. 
The second if logic is 'Look for elb's and update them if we have them', they dont need to exist

ashb commented on bootstrap_cfn/iam.py:
```diff
@@ -29,7 +31,57 @@ def upload_ssl_certificate(self, ssl_config, stack_name):
+        cert_arn = None
+
+        # Collect all certificates into a list
+        certs = self.conn_iam.get_all_server_certs()
```
Can you use `iam.get_server_certificate(cert_name)` instead? http://boto.readthedocs.org/en/latest/ref/iam.html#boto.iam.connection.IAMConnection.get_server_certificate

ashb commented on bootstrap_cfn/iam.py:
```diff
@@ -29,7 +31,57 @@ def upload_ssl_certificate(self, ssl_config, stack_name):
+        for cert in certlist:
+            if cert['server_certificate_name'] == cert_name:
+                cert_arn = cert.arn
+                break
```
When we find the result it is cleaner to just return the result directly rather than break and check again if we've found an answer. (thought this loop might go away) if we can use get_server_certificate directly.

ashb commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -317,3 +319,26 @@ def rsync():
         use_sudo=True)
+
+@task 
+def update_certs():
```
If possible it would be really nice if this method was idempontent - i.e. if the certs haven't changed then running this function should do nothing.

I'm not sure how feasible that is to do thought...

ashb commented:
Can you also rebase this into a small number of commits? In this case probably just 1.

Commit:
[enable_cert_updates]
    Created a fab task to update certificates on a stack. Checks are carried out to see if the current local and remote certificates are the same,
    if not the certificates are uploaded and and load balancers with certificates in the stack will be updated
    fab_tasks:
        - Create a fab task update_certs to update the ssl certificates from the config
        - Small delay between updating certificates on instances and load balancer to avoid setting the load balancer cert when ARN is available but the
            certificate is not properly registered
    iam:
        - Added getting an arn from a certificate name to the IAM class.
        - Simplify getting an arn for certs using boto get_server_certificate
        - Handle exceptions when trying to delete/upload a certificate
        - Simplify upload ssl logic to error on trying to update a missing
            certificate
    elb:
        - Create an ELB class to handle interaction with the stacks load balancers.
        - Throw an exception if we try to set certs on elbs when none are defined in the config
    cloudformation:
        - Added utility method to get a list of load balancers for the stack
    test_iam:
        - Added test cases to cover noew iam things

ashb commented:
@niallcreech Travis is reporting that the tests are failing

```
FAIL: test_ssl_delete (tests.test.CfnTestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/ministryofjustice/bootstrap-cfn/tests/test.py", line 246, in test_ssl_delete
    self.assertTrue(x)
AssertionError: False is not true
```

Did you change the API here perhaps?

sevenmachines commented:
Yep, I'm on it. its not the API, delete checks to see if the cert exists. I'm setting up the test to mock the call to list_server_certs.

sevenmachines commented:
- Update test_ssl_delete test case to reflect new 'check cert exists' flow for IAM class

Now expanding tests to cover update_certs fab_task

sevenmachines commented:
Need to fill out the test cases then hopefully done

sevenmachines commented:
Added more test cases, hopefully ready for review

mattmb commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -312,3 +314,45 @@ def rsync():
+    Also handle settings the certificates on ELB's
+    """
+    if verbose:
+        logging.basicConfig(level=logging.INFO)
```
Rather than setting a flag called verbose here, it would be better to just use the logger with different levels. Like `logging.info` when you want it to appear for everyone and `logging.debug` when you want it only if the verbosity is increased. Then we can set the level globally in the fabfile maybe?

mattmb commented on bootstrap_cfn/elb.py:
```diff
@@ -0,0 +1,80 @@
+            load_balancer_resources = self.cfn.get_stack_load_balancers(stack_name)
+            found_load_balancer_names = []
+            for load_balancer_resource in load_balancer_resources:
+                found_load_balancer_names.append(load_balancer_resource.physical_resource_id)
```
This can be made a bit more pythony like this:

```
found_load_balancer_names = [ lb.physical_resource_id for lb in load_balancer_resources]
```

mattmb commented on bootstrap_cfn/elb.py:
```diff
@@ -0,0 +1,80 @@
+
+            # Use load balancer names to filter gettingload balancer details
+            if len(found_load_balancer_names) > 0:
+                load_balancers = self.conn_elb.get_all_load_balancers(load_balancer_names=found_load_balancer_names)
```
Do you know what get_all_load_balancers returns if there are none found? If it's an empty list (or iterator) then we might not need the `if len()` on line 58.

Likewise if you call it with load_balancer_names as empty it might just return an empty list and you wont need the one on 53 either.

mattmb replied:
Note this isn't super important, just if you happen to know how it works.

mattmb replied:
Just noticed you're using it to throw an exception so ignore me :-)

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+        private_key = ssl_data['key']
+        try:
+            cert_chain = ssl_data['chain']
+        except:
```
This should be `except KeyError` (always good to catch exceptions as specifically as possible.

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+        Returns:
+            success(bool): True if certificate is uploaded, False otherwise
+        """
+        success = False
```
You can do away with this by just doing `return False` or `return True`

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+                                                                   stack_name,
+                                                                   ssl_data)
+                if cert_exists:
+                    if force:
```
`if cert_exists and force`

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+                    upload_success = self.upload_certificate(cert_name,
+                                                             stack_name,
+                                                             ssl_data,
+                                                             force)
```
So we're doing a delete and create? Does this work when the certificate is already attached to an ELB? What happens in the small time in between when there's no cert?

mattmb replied:
Oh I see, it's ok because the cert your deleting remains attached to the listener on the ELB?

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+            exists(bool): True if remote AWS certificate exists, false otherwise
+        """
+
+        exists = False
```
likewise with this, clearer to just `return False`

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+             "cert": remote_cert_body,
+             "chain": remote_cert_chain,
+             "key": remote_cert_key,
+             }
```
Can you make this into one block like:
`"cert": remote_cert_certificate.get("certificate_body", None),` etc

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+             "key": remote_cert_key,
+             }
+            # Compare the local cert and chain certificates to remote
+            if self.compare_certificate_data(remote_cert_data, remote_cert_data):
```
This looks like a mistake?

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,299 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+
+        return are_equal
+
+    def compare_text_between_strings(self,
```
Can we put the start text/end text in this method (rather than passing it in). Think you might be trying to keep it generic but in this case I feel like it would be better to make it specific to the problem. Maybe rename it to something like compare_certs_body or something like that.

mattmb commented:
I still need to review the tests. :smile:

mattmb commented on bootstrap_cfn/fab_tasks.py:
```diff
@@ -312,3 +314,45 @@ def rsync():
+    # Arbitrary wait to allow SSL upload to register with AWS
+    # Otherwise, we can get an ARN for the load balancer certificates
+    # without it being ready to assign
+    time.sleep(3)
```
Ugh. Is that a limitation of the AWS API? If so amazon suck :angry:

sevenmachines replied:
seems to be, couldn't find another explanation for the delay in availablity

sevenmachines commented:
- Dropped verbose arg from fab task
- elb: Made load_balance search more pythonistic
- iam: Made exception catching more specific to KeyError
- iam: Move to pythonic mid-method returns rather than collecting
  results and returning on method exit
- iam: Combine cert_exists and force checks on if
- iam: combine some dictionary calls to compact code
- iam: Fix remote and local cert data comparison
- iam: Compact update_ssl_certificates checks on force and exists

sevenmachines commented:
Not sure about some of this 'forcing' logic

mattmb commented on bootstrap_cfn/elb.py:
```diff
@@ -0,0 +1,77 @@
+            load_balancer_resources = self.cfn.get_stack_load_balancers(stack_name)
+            found_load_balancer_names = [lb.physical_resource_id for lb in load_balancer_resources]
+            # Use load balancer names to filter gettingload balancer details
+            if len(found_load_balancer_names) > 0:
```
think we might want `load_balancers=[]` here or in the else block because if this is not true the next if statement will fail (if I'm reading it right).

mattmb commented on bootstrap_cfn/iam.py:
```diff
@@ -18,18 +20,292 @@ def __init__(self, aws_profile_name, aws_region_name='eu-west-1'):
+                                                  cert_data1.get("cert", None),
+                                                  cert_data2.get("cert", None))
+        if not certs_are_equal:
+            logging
```
missing the log message here?

mattmb commented:
@niallcreech and I have just had a good chat about this.

Two things we thought of:
- Force may not be necessary. Use cases are my repo has the correct cert, I run the tool, nothing changes. My repo has a new cert, I run the tool, the cert gets updated.
- Multiple ELBs with different certs. We think it's probably safe the way it works at the moment because the cert wont be updated, the ARN wont change etc. But might be nicer to track what's changing (rather than an updated count (maybe a dictionary of what's updated)

stuartornum commented:
This is a massive PR, I have no idea where to start

sevenmachines commented:
Mostly covered with Matt yesterday, I’ve got a couple of small changes to make but unfortunately I’ve no internet until the afternoon. Press on with any bootstrap work you’re doing, I’ll fix this up around it once you’re finished

sevenmachines commented:
- Reworked IAM force options to remove them and have sane logic
- Return list of updated elb's when we run update_ssl's
- Updated tests to reflect new force and list return logic

sevenmachines commented:
Will need to wait until trim_the_fat has been merged and see how rebasey this will be before moving to review

mattmb commented:
Cool, thanks @niallcreech I've got my fingers crossed with the git gods that it won't be too conflicty :smile:

mattmb commented:
I'd say this looks good now. Anyone else want to comment before I merge?
